math/minuit2/src/NumericalDerivator.cxx

```diff
@@ -0,0 +1,336 @@
+// @(#)root/mathcore:$Id$
+// Authors: L. Moneta, J.T. Offermann, E.G.P. Bos    2013-2018
+//
+/**********************************************************************
+ *                                                                    *
+ * Copyright (c) 2013 , LCG ROOT MathLib Team                         *
+ *                                                                    *
+ **********************************************************************/
+/*
+ * NumericalDerivator.cxx
+ *
+ *  Original version created on: Aug 14, 2013
+ *      Authors: L. Moneta, J. T. Offermann
+ *  Modified version created on: Sep 27, 2017
+ *      Author: E. G. P. Bos
+ *
+ *      NumericalDerivator was essentially a slightly modified copy of code
+ *      written by M. Winkler, F. James, L. Moneta, and A. Zsenei for Minuit2,
+ *      Copyright (c) 2005 LCG ROOT Math team, CERN/PH-SFT. Original version:
+ *      https://github.com/lmoneta/root/blob/lvmini/math/mathcore/src/NumericalDerivator.cxx
+ *
+ *      This class attempts to more closely follow the Minuit2 implementation.
+ *      Modified things (w.r.t. original) are indicated by MODIFIED.
+ */
+
+#include "Minuit2/NumericalDerivator.h"
+#include <cmath>
+#include <algorithm>
+#include <Math/IFunction.h>
+#include <iostream>
+#include <TMath.h>
+#include <cassert>
+#include "Fit/ParameterSettings.h"
+
+#include <Math/Minimizer.h> // needed here because in Fitter is only a forward declaration
+
+namespace ROOT {
+namespace Minuit2 {
+
+NumericalDerivator::NumericalDerivator(bool always_exactly_mimic_minuit2)
+   : fAlwaysExactlyMimicMinuit2(always_exactly_mimic_minuit2)
+{
+}
+
+NumericalDerivator::NumericalDerivator(double step_tolerance, double grad_tolerance, unsigned int ncycles,
+                                       double error_level, bool always_exactly_mimic_minuit2)
+   : fStepTolerance(step_tolerance), fGradTolerance(grad_tolerance), fNCycles(ncycles), fUp(error_level),
+     fAlwaysExactlyMimicMinuit2(always_exactly_mimic_minuit2)
+{
+}
+
+// deep copy constructor
+NumericalDerivator::NumericalDerivator(const NumericalDerivator &other)
+   : fStepTolerance(other.fStepTolerance), fGradTolerance(other.fGradTolerance), fNCycles(other.fNCycles),
+     fUp(other.fUp), fVal(other.fVal), fVx(other.fVx), fVxExternal(other.fVxExternal), fDfmin(other.fDfmin),
+     fVrysml(other.fVrysml), fPrecision(other.fPrecision), fAlwaysExactlyMimicMinuit2(other.fAlwaysExactlyMimicMinuit2),
+     fVxFValCache(other.fVxFValCache)
+{
+}
+
+void NumericalDerivator::SetStepTolerance(double value)
+{
+   fStepTolerance = value;
+}
+
+void NumericalDerivator::SetGradTolerance(double value)
+{
+   fGradTolerance = value;
+}
+
+void NumericalDerivator::SetNCycles(unsigned int value)
+{
+   fNCycles = value;
+}
+
+void NumericalDerivator::SetErrorLevel(double value)
+{
+   fUp = value;
+}
```

Comment:
These might be inlined, and one-line `///` comments could be added to help users use this class.

Reply:
Maybe @lmoneta or you or someone else could later on add some helpful comments. The algorithm here is a replica of the `Numerical2PGradientCalculator`. I understand more or less how it works, but not necessarily why all parameters are the way they are. Someone like @lmoneta might be better able to do this.